File: 0131-palindrome-partitioning/0131-palindrome-partitioning.cpp
class Solution {
public:

    bool isPalindrome(string s){
        int l=0, r=s.size()-1;
        while(l<r) if(s[l++]!=s[r--]) return false;
        return true;
    }
    void helper(string s, int i, vector<vector<string>> &ans, vector<string> &temp){
        if(i==s.size()){
            ans.push_back(temp);
            return ;
        }
        for(int j=i; j<s.length(); j++){
            string t= s.substr(i,j-i+1);
            if(isPalindrome(t)){
                temp.push_back(t);
                helper(s,j+1,ans, temp);
                temp.pop_back();
            }
        }
    }
    vector<vector<string>> partition(string s) {
        vector<vector<string>> ans;
        vector<string> temp;
        helper(s,0,ans, temp);
        return ans;
    }
};

File: 0131-palindrome-partitioning/0131-palindrome-partitioning.java
class Solution {
    boolean isPal(String x){
        int l=0, r= x.length()-1;
        while(l<r){
            if(x.charAt(l)!= x.charAt(r)) return false;
            l++;
            r--;
        }
        return true;
    }
    List<List<String>> res;
    void helper(String s, int idx, List<String> temp){
        if(idx>=s.length()){
            res.add(new ArrayList<>(temp));
            return;
        }
        String x="";
        for(int i=idx; i<s.length(); i++){
            x+=s.charAt(i);
            if(isPal(x)){
                temp.add(x);
                helper(s, i+1, temp);
                temp.remove(temp.size()-1);
            }
        }
        return ;
    }
    public List<List<String>> partition(String s) {
        res= new ArrayList<>();
        List<String> temp= new ArrayList<String>();
        helper(s, 0, temp);
        return res;
    }
}